### ISSUE TITLE
OSSL_PROPERTY_LIST concurrency

### ISSUE THREAD

**Issue by jwalch** [openssl/openssl#18613]
I am trying to understand the thread-safety model in evp_set_parsed_default_properties(), in particular around this:

```
    OSSL_PROPERTY_LIST **plp = ossl_ctx_global_properties(libctx, loadconfig);
[...]
        ossl_property_free(*plp);
        *plp = def_prop;
```

As best I can tell, the intent is for this to be lock-less with respect to the OSSL_PROPERTY_LIST usage in this particular function. Otherwise it is hard to see how concurrent EVP_set_default_properties() & EVP_default_properties_is_fips_enabled() would be safe. (There doesn't seem to be any indication at https://www.openssl.org/docs/man3.0/man3/EVP_default_properties_enable_fips.html that they AREN'T.)

Assuming that's the case, isn't there room for a UAF in a different thread between the last two lines above? Shouldn't we 
write the new pointer value (let's assume this is safely atomic for now) prior to freeing memory at the value of the old one?

It SEEMS like the correct thing to do is atomically swap def_prop and *plp and then free def_prop (ie, original *plp). But there is an extremely good chance that I'm misunderstanding the concurrency model here...

**Comment by t8m** [comment]
Setting the default properties is inherently thread unsafe because setting them involves potentially enabling/disabling algorithms and/or implementations and doing that concurrently with other threads that use the same library context to do some operations does not make much sense.

Proper usage of default properties IMO involves calling EVP_set_default_properties() only during the libcontext initialization. Alternatively the application could do its own locking if it is able to cope with algorithm implementations availability changing as the default properties are changed.

**Comment by paulidale** [comment]
Agreed with @t8m.  This function was never meant to be thread safe.

If it is essential, we could make it safe.

**Comment by t8m** [comment]
Marking it as a documentation issue. IMO making it thread safe would be a feature -> for master branch only.

**Comment by paulidale** [comment]
The idea has always been to set the default query up up front and live with it for the duration.
